### PR TITLE
Tests/benchmarks

### PR DESCRIPTION
**Pull request by benjamin-hejl**
A lot of small changes throughout the code base in order.
 - refactored ISolver.Solve to distinct Undecidable from Unsat
 - fixed a lot of small bugs
 - created a utility for benchmarking exploration (project `dnWalker.Benchmarks`). The script run-benchmarks.ps1 runs this utility using Examples assembly and a collection of 17 methods from it (used in the thesis)
   - generates tests which can be run in the project Examples.Tests.Generated and observe coverage
   - generates exploration data together with cfg and constraint tree .dot files located in the `benchmark-exploration/{class_under_test}_{method_under_test}/` directory
   - statistics are written to `benchmark-exploration/stats.csv` and on the console output
   - takes number of runs and warmup runs as arguments, the script runs 10 times with 5 warmup runs
   - warmup uses the first method in provided list

### CODE REVIEW

**Comment by kfrajtak** [comment]
Ta nova strategie je to nejdulezitejsi, ta mela byt v jednom PR a zbytek v dalsim

**Comment by benjamin-hejl** [comment]
> vyextrahoval jsem sem: (sám o sobě ale nepůjde zkompilovat)
https://github.com/kfrajtak/dnWalker/pull/96

File: dnWalker.Benchmarks/ConsoleStatsOutputHelper.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace dnWalker.Benchmarks
+{
+    public class ConsoleStatsOutputHelper
+    {
+        private static readonly ConsoleTableOptions TableOptions = new ConsoleTableOptions()
+        {
+            Columns = new[] { "Method", "Expl. Time [ms]", "Gen. Time [ms]", "OK", "ERROR", "DON'T KNOW", "Tests" },
```
Comment (kfrajtak) [inline dnWalker.Benchmarks/ConsoleStatsOutputHelper.cs:15]: na benchmarky existuje hezka knihovna

File: dnWalker.Explorations/ConcolicExploration.cs [hunk 2]
```diff
@@ -9,7 +9,7 @@ public partial class ConcolicExploration
     {
         private readonly List<ConcolicExplorationIteration> _iterations = new List<ConcolicExplorationIteration>();
 
-        private ConcolicExploration(string assemblyName, string assemblyFileName, IMethod methodUnderTest, string solver, DateTime start, DateTime end, bool failed)
+        private ConcolicExploration(string assemblyName, string assemblyFileName, IMethod methodUnderTest, string solver, string strategy, DateTime start, DateTime end, bool failed)
```
Comment (kfrajtak) [inline dnWalker.Explorations/ConcolicExploration.cs:12]: vsechno se to muze dat do tridy `ConcolicExplorationParams` a nebude se to muset pokazde menit

**Review by kfrajtak — COMMENTED** [review]
kde je ta nova strategie?

**Comment by benjamin-hejl** [comment]
> kde je ta nova strategie?
v PR https://github.com/kfrajtak/dnWalker/pull/96

**Comment by kfrajtak** [comment]
ok, tohle zmergujte

File: dnWalker.Benchmarks/ConsoleStatsOutputHelper.cs (hunk 1 above)
Comment (benjamin-hejl) [reply dnWalker.Benchmarks/ConsoleStatsOutputHelper.cs:15]: já vím, ale zároveň jsem chtěl mít dostatečnou kontrolu nad artefakty (vygenerované testy, cfg, constraint tree grafy atp) a možná by mi to umožňovala, ale tohle bylo rychlejší, než se prohrabat v dokumentaci...

**Pull request merged**